common/authorization/oauthAuthorizer.go

```diff
@@ -0,0 +1,108 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package authorization
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/cristalhq/jwt/v3"
+	"go.uber.org/yarpc"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/config"
+)
+
+type oauthAuthority struct {
+	authorizationCfg config.OAuthAuthorizer
+}
+
+type jwtClaims struct {
+	Sub        string
+	Name       string
+	Permission string
+	Domain     string
+	Iat        int64
+}
+
+// NewOAuthAuthorizer creates a no-op authority
+func NewOAuthAuthorizer(
+	authorizationCfg config.OAuthAuthorizer,
+) Authorizer {
+	return &oauthAuthority{
+		authorizationCfg: authorizationCfg,
+	}
+}
+
+func (a *oauthAuthority) Authorize(
+	ctx context.Context,
+	attributes *Attributes,
+) (Result, error) {
+	call := yarpc.CallFromContext(ctx)
+	token := call.Header(common.AuthorizationTokenHeaderName)
+	// parseTokenAndVerify could either return the claims or a bool. I'm returning the claims currently
+	// because we might use it to populate values in the context
+	_, err := a.parseTokenAndVerify(token, attributes)
+	if err != nil {
+		return Result{Decision: DecisionDeny}, err
+	}
+	return Result{Decision: DecisionAllow}, nil
+}
+
+func (a *oauthAuthority) parseTokenAndVerify(tokenStr string, attributes *Attributes) (*jwtClaims, error) {
+	publicKey, err := common.StringToRSAPublicKey(a.authorizationCfg.JwtCredentials.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	algorithm := jwt.Algorithm(a.authorizationCfg.JwtCredentials.Algorithm)
+	verifier, err := jwt.NewVerifierRS(algorithm, publicKey)
+	if err != nil {
+		return nil, err
+	}
+	token, verifyErr := jwt.ParseAndVerifyString(tokenStr, verifier)
+	if verifyErr != nil {
+		return nil, verifyErr
+	}
+	var claims jwtClaims
+	_ = json.Unmarshal(token.RawClaims(), &claims)
+	validationErr := a.validateClaims(claims, attributes)
+	if validationErr != nil {
+		return nil, validationErr
+	}
+
+	return &claims, nil
+}
+
+func (a *oauthAuthority) validateClaims(claims jwtClaims, attributes *Attributes) error {
+	if claims.Iat < time.Now().Unix() {
```

Comment:
Iat is not expiring time. So it should be 
```
if claims.Iat.Add(claim.TTL) < time.Now().Unix() {
```
BTW can you add TTL into the claim? I just updated the proposal. 
That will be more flexible to control the expiring. Some token may need longer TTL(like users CLI because refreshing every hour is annoying) but some can be shorter(like worker can refresh it automatically).